json/src/ser.rs

```diff
@@ -307,6 +308,50 @@ impl<'a, W, F> ser::Serializer for &'a mut Serializer<W, F>
         try!(self.formatter.begin_object_value(&mut self.writer));
         self.serialize_map(Some(len))
     }
+
+    fn collect_str<T>(self, value: &T) -> Result<Self::Ok>
+        where T: fmt::Display,
+    {
+        use std::fmt::Write;
+
+        struct Adapter<'ser, W: 'ser, F: 'ser> {
+            writer: &'ser mut W,
+            formatter: &'ser mut F,
+            error: Option<Error>,
+        }
+
+        impl<'ser, W, F> Write for Adapter<'ser, W, F>
+            where W: io::Write,
+                  F: Formatter,
+        {
+            fn write_str(&mut self, s: &str) -> fmt::Result {
+                assert!(self.error.is_none());
+                match self.formatter.write_string_fragment(&mut self.writer, s.as_bytes()) {
```

Comment:
Made `write_string_fragment` take a `&str` and contain the whole loop from the function `format_escaped_str`.